### PR TITLE
Add qos option to override qos

### PR DESCRIPTION
**Pull request by wep21**
I added qos option to override qos via ros parameter.

### CODE REVIEW

**Comment by wep21** [comment]
@mjcarroll could you review this?

**Comment by abstanton** [comment]
any update on this?

**Comment by wep21** [comment]
@gbiggs friendly ping

**Comment by gbiggs** [comment]
Can you please add a test that verifies:

1. That when the parameter is not set, the original QoS is preserved.
2. That when the parameter is set, the specified QoS is correctly used.

**Comment by abstanton** [comment]
@gbiggs my team needs this quite badly, how could I take over the patch to add these tests?

**Comment by wep21** [comment]
I found out that subsriber plugin with options has been already implemented in #186.
I will take another look at publisher implementation and try to make some tests.

**Comment by wep21** [comment]
@gibiggs I added basic tests at https://github.com/ros-perception/image_common/pull/208/commits/908699407aa9fd1afe172fcf99717c02f58824ca.

File: image_transport/test/test_publisher.cpp [hunk 1]
```diff
@@ -46,31 +46,61 @@ class TestPublisher : public ::testing::Test
...
+  rclcpp::PublisherOptions options;
+  options.qos_overriding_options = rclcpp::QosOverridingOptions(
+  {
+    rclcpp::QosPolicyKind::Depth,
+    rclcpp::QosPolicyKind::Durability,
+    rclcpp::QosPolicyKind::History,
+    rclcpp::QosPolicyKind::Reliability,
+  });
+  auto pub = image_transport::create_publisher(
+    node_.get(), "camera/image", rmw_qos_profile_default, options);
```
Comment (gbiggs) [inline image_transport/test/test_publisher.cpp]: You should also test the case where `options` is not provided.

File: image_transport/test/test_subscriber.cpp [hunk 2]
```diff
@@ -66,6 +66,38 @@ TEST_F(TestPublisher, shutdown) {
...
+  {
+    rclcpp::QosPolicyKind::Depth,
+    rclcpp::QosPolicyKind::Durability,
+    rclcpp::QosPolicyKind::History,
+    rclcpp::QosPolicyKind::Reliability,
+  });
+  std::function<void(const sensor_msgs::msg::Image::ConstSharedPtr & msg)> fcn =
+    [](const auto & msg) {(void)msg;};
+  auto sub = image_transport::create_subscription(
+    node_.get(), "camera/image", fcn, "raw", rmw_qos_profile_default, options);
```
Comment (gbiggs) [inline image_transport/test/test_subscriber.cpp]: You should also test the case where `options` is not provided.

File: image_transport/include/image_transport/raw_publisher.hpp [hunk 3]
```diff
@@ -66,6 +66,17 @@ class RawPublisher : public SimplePublisherPlugin<sensor_msgs::msg::Image>
   {
     return base_topic;
   }
+
+  using PublisherPlugin::advertiseImpl;
```
Comment (gbiggs) [inline image_transport/include/image_transport/raw_publisher.hpp]: Why is this line necessary?

File: image_transport/include/image_transport/publisher_plugin.hpp [hunk 4]
```diff
@@ -136,6 +137,19 @@ class PublisherPlugin
   virtual void advertiseImpl(
     rclcpp::Node * nh, const std::string & base_topic,
     rmw_qos_profile_t custom_qos) = 0;
+
+  virtual void advertiseImpl(
```
Comment (gbiggs) [inline image_transport/include/image_transport/publisher_plugin.hpp:141]: I think it would be better to add `rclcpp::PublisherOptions options = rclcpp::PublisherOptions()` to the existing abstract `advertiseImpl` function. Then plugins can provide the same default options or customised options as they see fit and pass them along. That would avoid the need to overload the `advertiseImpl` function.

File: image_transport/include/image_transport/publisher_plugin.hpp [hunk 5]
```diff
@@ -136,6 +137,19 @@ class PublisherPlugin
...
+    rclcpp::Node * node,
+    const std::string & base_topic,
+    rmw_qos_profile_t custom_qos,
+    rclcpp::PublisherOptions options)
+  {
+    (void) options;
+    RCLCPP_ERROR(
+      node->get_logger(),
+      "PublisherPlugin::advertiseImpl with four arguments has not been overridden");
```
Comment (gbiggs) [inline image_transport/include/image_transport/publisher_plugin.hpp:150]: If it hasn't been overridden, why is it not pure abstract?

File: image_transport/test/test_publisher.cpp (hunk 1 above)
Comment (wep21) [reply image_transport/test/test_publisher.cpp]: @gbiggs add test at https://github.com/ros-perception/image_common/pull/208/commits/f6ab00a0db2be5bb79ef270a6c8d0dd70babc241

File: image_transport/test/test_subscriber.cpp (hunk 2 above)
Comment (wep21) [reply image_transport/test/test_subscriber.cpp]: @gbiggs add test at https://github.com/ros-perception/image_common/pull/208/commits/f6ab00a0db2be5bb79ef270a6c8d0dd70babc241

File: image_transport/include/image_transport/raw_publisher.hpp (hunk 3 above)
Comment (wep21) [reply image_transport/include/image_transport/raw_publisher.hpp]: remove the line at https://github.com/ros-perception/image_common/pull/208/commits/df25b2386124d7286518f64340a58cb28dc4aaf4

File: image_transport/include/image_transport/publisher_plugin.hpp (hunk 4 above)
Comment (wep21) [reply image_transport/include/image_transport/publisher_plugin.hpp:141]: I implement this in accordance with subscriber impl. I guess this is to keep the interface for the inherited classes in another project such as https://github.com/ros-perception/image_transport_plugins.
@audrow Please correct me if I'm wrong.

File: image_transport/test/test_publisher.cpp [hunk 6]
```diff
@@ -46,31 +46,87 @@ class TestPublisher : public ::testing::Test
...
-TEST_F(TestPublisher, Shutdown) {
+TEST_F(TestPublisher, shutdown) {
   auto pub = image_transport::create_publisher(node_.get(), "camera/image");
   EXPECT_EQ(node_->get_node_graph_interface()->count_publishers("camera/image"), 1u);
   pub.shutdown();
   EXPECT_EQ(node_->get_node_graph_interface()->count_publishers("camera/image"), 0u);
 }
 
+TEST_F(TestPublisher, qos_override) {
+  // without options
```
Comment (gbiggs) [inline image_transport/test/test_publisher.cpp]: You should make the with and without options separate test cases rather than recreating the initial conditions halfway through the test case.

File: image_transport/test/test_publisher.cpp [hunk 7]
```diff
@@ -46,31 +46,87 @@ class TestPublisher : public ::testing::Test
...
+  pub = image_transport::create_publisher(node_.get(), "camera/image", rmw_qos_profile_default);
+
+  endpoint_info_vec = node_->get_publishers_info_by_topic("camera/image");
+  EXPECT_EQ(
+    endpoint_info_vec[0].qos_profile().reliability(),
+    rclcpp::ReliabilityPolicy::Reliable);
+  pub.shutdown();
+
+  // with options
+  node_ = rclcpp::Node::make_shared("test_publisher");
```
Comment (gbiggs) [inline image_transport/test/test_publisher.cpp]: You're making an entire node here just to replace it later on on line 114. Line 114 should create a new node with a new variable name rather than overwriting the one from the test fixture. You could even consider making this a stand-alone test case rather than using the test fixture, since you don't actually use the node created by the test fixture.

File: image_transport/test/test_publisher.cpp (hunk 6 above)
Comment (wep21) [reply image_transport/test/test_publisher.cpp]: @gbiggs I separated with and without options test case at https://github.com/ros-perception/image_common/pull/208/commits/16b9d874c3c54378c82829e5754f7c1950ebb84a

File: image_transport/test/test_publisher.cpp (hunk 7 above)
Comment (wep21) [reply image_transport/test/test_publisher.cpp]: @gbiggs  I separated qos override test case at https://github.com/ros-perception/image_common/pull/208/commits/4eda95d6b38b6a2b33d9ac230ead47357b7be2b7.

**Comment by wep21** [comment]
@audrow Could you comment about @gbiggs 's review?

**Comment by gbiggs** [comment]
CI:

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/16650/)
- Linux-aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/11232/)
- Windows [Build Status](https://ci.ros2.org/job/ci_windows/17047/)

**Comment by gbiggs** [comment]
CI:

- Linux [Build Status](https://ci.ros2.org/job/ci_linux/16651/)
- Linux-aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/11233/)
- Windows [Build Status](https://ci.ros2.org/job/ci_windows/17048/)

**Comment by gbiggs** [comment]
CI has passed so merging this. Thank you for the contribution!

**Comment by gecastro** [comment]
Is there a plan to backport this feature to galactic?

**Comment by gbiggs** [comment]
Unfortunately not, as this would break API in Galactic.
